{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "from typing import Tuple\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "from source import load_avenio_files, categorical_columns_to_lower\n",
    "from transform import combine_tsv_files, load_process_and_store_spreadsheets, merge_mutation_spreadsheet_t0_with_t1, clean_mutation_columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# First make a spreadsheet with merged t0 and t1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "spread_sheet_filename= \"variant_list_20200409.xlsx\"\n",
    "spss_filename= \"clinical_20200419.sav\"\n",
    "columns = [\n",
    "    \"Allele Fraction\",\n",
    "    \"No. Mutant Molecules per mL\",\n",
    "    \"CNV Score\",\n",
    "]\n",
    "patient_mutations, phenotypes = load_avenio_files(\n",
    "    spread_sheet_filename, spss_filename\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>No Patient</th>\n",
       "      <th>ctDNA T0</th>\n",
       "      <th>ctDNA T1</th>\n",
       "      <th>PFS</th>\n",
       "      <th>PFS censor</th>\n",
       "      <th>OS</th>\n",
       "      <th>OS censor</th>\n",
       "      <th>No. Mutant Molecules per mL</th>\n",
       "      <th>Variant Depth</th>\n",
       "      <th>Unique Depth</th>\n",
       "      <th>CNV Score</th>\n",
       "      <th>Unnamed: 30</th>\n",
       "      <th>Unnamed: 31</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>852</th>\n",
       "      <td>852.0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>284</td>\n",
       "      <td>4</td>\n",
       "      <td>852</td>\n",
       "      <td>4</td>\n",
       "      <td>66.30</td>\n",
       "      <td>25.0</td>\n",
       "      <td>14482.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>951</th>\n",
       "      <td>951.0</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>1242</td>\n",
       "      <td>18</td>\n",
       "      <td>1422</td>\n",
       "      <td>18</td>\n",
       "      <td>6642.85</td>\n",
       "      <td>4289.0</td>\n",
       "      <td>73893.0</td>\n",
       "      <td>26.825325</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>952</th>\n",
       "      <td>952.0</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>112</td>\n",
       "      <td>16</td>\n",
       "      <td>928</td>\n",
       "      <td>16</td>\n",
       "      <td>157890.00</td>\n",
       "      <td>5650.0</td>\n",
       "      <td>39763.0</td>\n",
       "      <td>93.196091</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>957</th>\n",
       "      <td>957.0</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>484</td>\n",
       "      <td>22</td>\n",
       "      <td>1034</td>\n",
       "      <td>22</td>\n",
       "      <td>130186.40</td>\n",
       "      <td>8999.0</td>\n",
       "      <td>205583.0</td>\n",
       "      <td>40.141279</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962</th>\n",
       "      <td>962.0</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>10344</td>\n",
       "      <td>0</td>\n",
       "      <td>10344</td>\n",
       "      <td>0</td>\n",
       "      <td>453.70</td>\n",
       "      <td>191.0</td>\n",
       "      <td>13029.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7183</th>\n",
       "      <td>7183.0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>1092</td>\n",
       "      <td>12</td>\n",
       "      <td>1908</td>\n",
       "      <td>12</td>\n",
       "      <td>597.40</td>\n",
       "      <td>318.0</td>\n",
       "      <td>77588.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7184</th>\n",
       "      <td>7184.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>721</td>\n",
       "      <td>0</td>\n",
       "      <td>721</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7185</th>\n",
       "      <td>7185.0</td>\n",
       "      <td>14</td>\n",
       "      <td>14</td>\n",
       "      <td>602</td>\n",
       "      <td>14</td>\n",
       "      <td>2016</td>\n",
       "      <td>14</td>\n",
       "      <td>595.50</td>\n",
       "      <td>295.0</td>\n",
       "      <td>88445.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7186</th>\n",
       "      <td>7186.0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "      <td>516</td>\n",
       "      <td>12</td>\n",
       "      <td>4056</td>\n",
       "      <td>12</td>\n",
       "      <td>5368.42</td>\n",
       "      <td>3550.0</td>\n",
       "      <td>39450.0</td>\n",
       "      <td>41.351371</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7187</th>\n",
       "      <td>7187.0</td>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>943</td>\n",
       "      <td>23</td>\n",
       "      <td>1012</td>\n",
       "      <td>23</td>\n",
       "      <td>7208.80</td>\n",
       "      <td>2068.0</td>\n",
       "      <td>159469.0</td>\n",
       "      <td>6.962640</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>174 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            No Patient  ctDNA T0  ctDNA T1    PFS  PFS censor     OS  \\\n",
       "Patient ID                                                             \n",
       "852              852.0         4         4    284           4    852   \n",
       "951              951.0        18        18   1242          18   1422   \n",
       "952              952.0        16        16    112          16    928   \n",
       "957              957.0        22        22    484          22   1034   \n",
       "962              962.0         6         6  10344           0  10344   \n",
       "...                ...       ...       ...    ...         ...    ...   \n",
       "7183            7183.0        12        12   1092          12   1908   \n",
       "7184            7184.0         0         0    721           0    721   \n",
       "7185            7185.0        14        14    602          14   2016   \n",
       "7186            7186.0        12        12    516          12   4056   \n",
       "7187            7187.0        23        23    943          23   1012   \n",
       "\n",
       "            OS censor  No. Mutant Molecules per mL  Variant Depth  \\\n",
       "Patient ID                                                          \n",
       "852                 4                        66.30           25.0   \n",
       "951                18                      6642.85         4289.0   \n",
       "952                16                    157890.00         5650.0   \n",
       "957                22                    130186.40         8999.0   \n",
       "962                 0                       453.70          191.0   \n",
       "...               ...                          ...            ...   \n",
       "7183               12                       597.40          318.0   \n",
       "7184                0                         0.00            0.0   \n",
       "7185               14                       595.50          295.0   \n",
       "7186               12                      5368.42         3550.0   \n",
       "7187               23                      7208.80         2068.0   \n",
       "\n",
       "            Unique Depth  CNV Score  Unnamed: 30  Unnamed: 31  \n",
       "Patient ID                                                     \n",
       "852              14482.0   0.000000          0.0          0.0  \n",
       "951              73893.0  26.825325          0.0          0.0  \n",
       "952              39763.0  93.196091          0.0          0.0  \n",
       "957             205583.0  40.141279          0.0          0.0  \n",
       "962              13029.0   0.000000          0.0          0.0  \n",
       "...                  ...        ...          ...          ...  \n",
       "7183             77588.0   0.000000          0.0          0.0  \n",
       "7184                 0.0   0.000000          0.0          0.0  \n",
       "7185             88445.0   0.000000          0.0          0.0  \n",
       "7186             39450.0  41.351371          0.0          0.0  \n",
       "7187            159469.0   6.962640          0.0          0.0  \n",
       "\n",
       "[174 rows x 13 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_mutations.groupby(['Patient ID']).agg(np.sum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> \u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m(274)\u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    273 \u001b[0;31m    \u001b[0;31m# Make a different document for each of the column pairs.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 274 \u001b[0;31m    \u001b[0;32mfor\u001b[0m \u001b[0mcolumn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    275 \u001b[0;31m        \u001b[0mcolumn_pair\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34mf\"T0: {column}\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34mf\"T1: {column}\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "ipdb> n\n",
      "> \u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m(275)\u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    274 \u001b[0;31m    \u001b[0;32mfor\u001b[0m \u001b[0mcolumn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 275 \u001b[0;31m        \u001b[0mcolumn_pair\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34mf\"T0: {column}\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34mf\"T1: {column}\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    276 \u001b[0;31m        \u001b[0;31m# Filter out the column pairs, and remove empty fields.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "ipdb> n\n",
      "> \u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m(277)\u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    276 \u001b[0;31m        \u001b[0;31m# Filter out the column pairs, and remove empty fields.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 277 \u001b[0;31m        \u001b[0mmutations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspread_sheet\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcolumn_pair\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"all\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfillna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    278 \u001b[0;31m        \u001b[0;31m# Repair dirty cells (with per cent signs etc.).\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "ipdb> n\n",
      "> \u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m(279)\u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    278 \u001b[0;31m        \u001b[0;31m# Repair dirty cells (with per cent signs etc.).\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 279 \u001b[0;31m        \u001b[0mclean_mutation_sheet\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclean_and_verify_data_frame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutations\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn_pair\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    280 \u001b[0;31m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "ipdb> p mutations\n",
      "     Patient ID    Gene Coding Change T0: Allele Fraction T1: Allele Fraction\n",
      "0           852    KRAS       c.34G>T              0.0018              0.0022\n",
      "1           852    PMS2     c.1394A>G              0.0013              0.0016\n",
      "2           951    BRAF      c.428A>G              0.0006              0.0013\n",
      "3           951    FLT4      c.844C>A              0.1208              0.0574\n",
      "4           951    FLT4      c.903C>A              0.1555              0.0722\n",
      "..          ...     ...           ...                 ...                 ...\n",
      "826        7187  NFE2L2       c.71G>C              0.0234              0.0253\n",
      "827        7187  NFE2L2       c.86A>G              0.0187              0.0025\n",
      "828        7187   PTCH1     c.3191C>T                   0              0.0013\n",
      "829        7187    PTEN      c.733C>T              0.0221              0.0228\n",
      "830        7187    ROS1     c.6288G>C              0.0099              0.0079\n",
      "\n",
      "[831 rows x 5 columns]\n",
      "ipdb> l\n",
      "\u001b[1;32m    274 \u001b[0m    \u001b[0;32mfor\u001b[0m \u001b[0mcolumn\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    275 \u001b[0m        \u001b[0mcolumn_pair\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34mf\"T0: {column}\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34mf\"T1: {column}\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    276 \u001b[0m        \u001b[0;31m# Filter out the column pairs, and remove empty fields.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    277 \u001b[0m        \u001b[0mmutations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspread_sheet\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcolumn_pair\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"all\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfillna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    278 \u001b[0m        \u001b[0;31m# Repair dirty cells (with per cent signs etc.).\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m--> 279 \u001b[0;31m        \u001b[0mclean_mutation_sheet\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclean_and_verify_data_frame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutations\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn_pair\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[1;32m    280 \u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    281 \u001b[0m        \u001b[0;31m# Perform `transformation` on the columns.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    282 \u001b[0m        mutations_transposed = transform_column_pair(\n",
      "\u001b[1;32m    283 \u001b[0m            \u001b[0mclean_mutation_sheet\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    284 \u001b[0m            \u001b[0;31m# Vocabulary is the entire dataset.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\n",
      "ipdb> n\n",
      "> \u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m(282)\u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    281 \u001b[0;31m        \u001b[0;31m# Perform `transformation` on the columns.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 282 \u001b[0;31m        mutations_transposed = transform_column_pair(\n",
      "\u001b[0m\u001b[0;32m    283 \u001b[0;31m            \u001b[0mclean_mutation_sheet\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "ipdb> p clean_mutation_columns\n",
      "<function clean_mutation_columns at 0x7f6a2ff4a050>\n",
      "ipdb> p clean_mutation_sheet\n",
      "     Patient ID    Gene Coding Change  T0: Allele Fraction  \\\n",
      "0           852    KRAS       c.34G>T               0.0018   \n",
      "1           852    PMS2     c.1394A>G               0.0013   \n",
      "2           951    BRAF      c.428A>G               0.0006   \n",
      "3           951    FLT4      c.844C>A               0.1208   \n",
      "4           951    FLT4      c.903C>A               0.1555   \n",
      "..          ...     ...           ...                  ...   \n",
      "826        7187  NFE2L2       c.71G>C               0.0234   \n",
      "827        7187  NFE2L2       c.86A>G               0.0187   \n",
      "828        7187   PTCH1     c.3191C>T               0.0000   \n",
      "829        7187    PTEN      c.733C>T               0.0221   \n",
      "830        7187    ROS1     c.6288G>C               0.0099   \n",
      "\n",
      "     T1: Allele Fraction  \n",
      "0                 0.0022  \n",
      "1                 0.0016  \n",
      "2                 0.0013  \n",
      "3                 0.0574  \n",
      "4                 0.0722  \n",
      "..                   ...  \n",
      "826               0.0253  \n",
      "827               0.0025  \n",
      "828               0.0013  \n",
      "829               0.0228  \n",
      "830               0.0079  \n",
      "\n",
      "[831 rows x 5 columns]\n",
      "ipdb> l\n",
      "\u001b[1;32m    277 \u001b[0m        \u001b[0mmutations\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mspread_sheet\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcolumn_pair\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"all\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfillna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    278 \u001b[0m        \u001b[0;31m# Repair dirty cells (with per cent signs etc.).\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    279 \u001b[0m        \u001b[0mclean_mutation_sheet\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclean_and_verify_data_frame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmutations\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumn_pair\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    280 \u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    281 \u001b[0m        \u001b[0;31m# Perform `transformation` on the columns.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m--> 282 \u001b[0;31m        mutations_transposed = transform_column_pair(\n",
      "\u001b[0m\u001b[1;32m    283 \u001b[0m            \u001b[0mclean_mutation_sheet\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    284 \u001b[0m            \u001b[0;31m# Vocabulary is the entire dataset.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    285 \u001b[0m            \u001b[0mgene_vocabulary\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mclean_mutation_sheet\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Gene\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    286 \u001b[0m            \u001b[0mcolumn_pair\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcolumn_pair\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m    287 \u001b[0m            \u001b[0mtransformation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtransformation\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\n",
      "ipdb> n\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> \u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m(283)\u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    282 \u001b[0;31m        mutations_transposed = transform_column_pair(\n",
      "\u001b[0m\u001b[0;32m--> 283 \u001b[0;31m            \u001b[0mclean_mutation_sheet\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    284 \u001b[0;31m            \u001b[0;31m# Vocabulary is the entire dataset.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "ipdb> n\n",
      "> \u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m(285)\u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    284 \u001b[0;31m            \u001b[0;31m# Vocabulary is the entire dataset.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 285 \u001b[0;31m            \u001b[0mgene_vocabulary\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mclean_mutation_sheet\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Gene\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    286 \u001b[0;31m            \u001b[0mcolumn_pair\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcolumn_pair\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "ipdb> \n",
      "> \u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m(286)\u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    285 \u001b[0;31m            \u001b[0mgene_vocabulary\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mclean_mutation_sheet\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Gene\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 286 \u001b[0;31m            \u001b[0mcolumn_pair\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcolumn_pair\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    287 \u001b[0;31m            \u001b[0mtransformation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtransformation\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "ipdb> \n",
      "> \u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m(287)\u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    286 \u001b[0;31m            \u001b[0mcolumn_pair\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcolumn_pair\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 287 \u001b[0;31m            \u001b[0mtransformation\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtransformation\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    288 \u001b[0;31m        )\n",
      "\u001b[0m\n",
      "ipdb> \n",
      "> \u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m(291)\u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    290 \u001b[0;31m        \u001b[0;31m# Add the patients that are not in the mutation list.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 291 \u001b[0;31m        patient_no_mutations = set(phenotypes[\"studynumber\"]) - set(\n",
      "\u001b[0m\u001b[0;32m    292 \u001b[0;31m            \u001b[0mmutations_transposed\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n",
      "ipdb> p mutations_transposed\n",
      "          KRAS      PMS2      BRAF      FLT4       KDR    PDGFRA  TERT  \\\n",
      "852   0.000990  0.000717  0.000000  0.000000  0.000000  0.000000   0.0   \n",
      "951   0.000000  0.000000  0.000411  0.088217  0.066381  0.014482   0.0   \n",
      "952   0.087793  0.000000  0.000000  0.003545  0.000000  0.082967   0.0   \n",
      "957   0.000000  0.000840  0.000000  0.166967  0.000000  0.000000   0.0   \n",
      "962   0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   \n",
      "...        ...       ...       ...       ...       ...       ...   ...   \n",
      "7181  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   \n",
      "7183  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   \n",
      "7185  0.002798  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   \n",
      "7186  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   0.0   \n",
      "7187  0.000000  0.000000  0.000000  0.000000  0.005775  0.000000   0.0   \n",
      "\n",
      "          TP53      EGFR     KEAP1  ...    CTNNB1  MAP2K2  IDH1  GATA3  BRCA1  \\\n",
      "852   0.000000  0.000000  0.000000  ...  0.000000     0.0   0.0    0.0    0.0   \n",
      "951   0.040469  0.000000  0.000000  ...  0.000000     0.0   0.0    0.0    0.0   \n",
      "952   0.095101  0.001510  0.095281  ...  0.000000     0.0   0.0    0.0    0.0   \n",
      "957   0.167365  0.002388  0.000000  ...  0.000000     0.0   0.0    0.0    0.0   \n",
      "962   0.005881  0.000000  0.000000  ...  0.000000     0.0   0.0    0.0    0.0   \n",
      "...        ...       ...       ...  ...       ...     ...   ...    ...    ...   \n",
      "7181  0.000000  0.000000  0.000000  ...  0.000858     0.0   0.0    0.0    0.0   \n",
      "7183  0.006759  0.000000  0.000000  ...  0.000000     0.0   0.0    0.0    0.0   \n",
      "7185  0.003208  0.000000  0.000000  ...  0.000000     0.0   0.0    0.0    0.0   \n",
      "7186  0.000000  0.000000  0.001841  ...  0.000000     0.0   0.0    0.0    0.0   \n",
      "7187  0.000000  0.000000  0.000000  ...  0.000000     0.0   0.0    0.0    0.0   \n",
      "\n",
      "      PIK3R1  CSF1R  GNAQ  RAF1  ARAF  \n",
      "852      0.0    0.0   0.0   0.0   0.0  \n",
      "951      0.0    0.0   0.0   0.0   0.0  \n",
      "952      0.0    0.0   0.0   0.0   0.0  \n",
      "957      0.0    0.0   0.0   0.0   0.0  \n",
      "962      0.0    0.0   0.0   0.0   0.0  \n",
      "...      ...    ...   ...   ...   ...  \n",
      "7181     0.0    0.0   0.0   0.0   0.0  \n",
      "7183     0.0    0.0   0.0   0.0   0.0  \n",
      "7185     0.0    0.0   0.0   0.0   0.0  \n",
      "7186     0.0    0.0   0.0   0.0   0.0  \n",
      "7187     0.0    0.0   0.0   0.0   0.0  \n",
      "\n",
      "[159 rows x 66 columns]\n",
      "ipdb> q\n"
     ]
    },
    {
     "ename": "BdbQuit",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mBdbQuit\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-38f6abf27d74>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mall_filename_prefix\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"output/all__harmonic_mean_\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0mtrain_filename_prefix\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"output/train__harmonic_mean_\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m     \u001b[0mtest_filename_prefix\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"output/test__harmonic_mean_\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m )\n",
      "\u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m in \u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m(spread_sheet_filename, spss_filename, transformation, columns, all_filename_prefix, train_filename_prefix, test_filename_prefix)\u001b[0m\n\u001b[1;32m    289\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    290\u001b[0m         \u001b[0;31m# Add the patients that are not in the mutation list.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 291\u001b[0;31m         patient_no_mutations = set(phenotypes[\"studynumber\"]) - set(\n\u001b[0m\u001b[1;32m    292\u001b[0m             \u001b[0mmutations_transposed\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    293\u001b[0m         )\n",
      "\u001b[0;32m/home/hylke/workspace/avenio/transform.py\u001b[0m in \u001b[0;36mload_process_and_store_spreadsheets\u001b[0;34m(spread_sheet_filename, spss_filename, transformation, columns, all_filename_prefix, train_filename_prefix, test_filename_prefix)\u001b[0m\n\u001b[1;32m    289\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    290\u001b[0m         \u001b[0;31m# Add the patients that are not in the mutation list.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 291\u001b[0;31m         patient_no_mutations = set(phenotypes[\"studynumber\"]) - set(\n\u001b[0m\u001b[1;32m    292\u001b[0m             \u001b[0mmutations_transposed\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    293\u001b[0m         )\n",
      "\u001b[0;32m/usr/lib/python3.7/bdb.py\u001b[0m in \u001b[0;36mtrace_dispatch\u001b[0;34m(self, frame, event, arg)\u001b[0m\n\u001b[1;32m     86\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0;31m# None\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     87\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mevent\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'line'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 88\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdispatch_line\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mframe\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     89\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mevent\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'call'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     90\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdispatch_call\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mframe\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/lib/python3.7/bdb.py\u001b[0m in \u001b[0;36mdispatch_line\u001b[0;34m(self, frame)\u001b[0m\n\u001b[1;32m    111\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstop_here\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mframe\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbreak_here\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mframe\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    112\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0muser_line\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mframe\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 113\u001b[0;31m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mquitting\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;32mraise\u001b[0m \u001b[0mBdbQuit\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    114\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtrace_dispatch\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mBdbQuit\u001b[0m: "
     ]
    }
   ],
   "source": [
    "def harmonic_mean(t0, t1):\n",
    "    return t0 * t1 / (t0 + t1)\n",
    "\n",
    "load_process_and_store_spreadsheets(\n",
    "    transformation=harmonic_mean,\n",
    "    all_filename_prefix=\"output/all__harmonic_mean_\",\n",
    "    train_filename_prefix=\"output/train__harmonic_mean_\",\n",
    "    test_filename_prefix=\"output/test__harmonic_mean_\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def difference(t0, t1):\n",
    "    return t1 - t0\n",
    "\n",
    "load_process_and_store_spreadsheets(\n",
    "    transformation=difference,\n",
    "    all_filename_prefix=\"output/all__difference_\",\n",
    "    train_filename_prefix=\"output/train__difference_\",\n",
    "    test_filename_prefix=\"output/test__difference_\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def relative_difference(t0, t1):\n",
    "    return (t1-t0)/t0\n",
    "\n",
    "load_process_and_store_spreadsheets(\n",
    "    transformation=relative_difference,\n",
    "    all_filename_prefix=\"output/all__relative_difference_\",\n",
    "    train_filename_prefix=\"output/train__relative_difference_\",\n",
    "    test_filename_prefix=\"output/test__relative_difference_\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def up_or_down(t0, t1):\n",
    "    return np.sign(t1 - t0)\n",
    "\n",
    "\n",
    "load_process_and_store_spreadsheets(\n",
    "    transformation=up_or_down,\n",
    "    all_filename_prefix=\"output/all__sign_\",\n",
    "    train_filename_prefix=\"output/train__sign_\",\n",
    "    test_filename_prefix=\"output/test__sign_\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Harmonic mean genomic variable.\n",
    "X_hm_af, y_hm_af = combine_tsv_files(\n",
    "    \"output/all__harmonic_mean__Allele Fraction.tsv\",\n",
    "    \"output/all__harmonic_mean__CNV Score.tsv\",\n",
    ")\n",
    "\n",
    "X_hm_af.merge(\n",
    "    y_hm_af, \n",
    "    left_index=True, \n",
    "    right_index=True\n",
    ").to_excel('output/all__harmonic_mean__af_cnv.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Harmonic mean genomic variable.\n",
    "X_hm_molecules, y_hm_molecules = combine_tsv_files(\n",
    "    \"output/all__harmonic_mean__No. Mutant Molecules per mL.tsv\",\n",
    "    \"output/all__harmonic_mean__CNV Score.tsv\",\n",
    ")\n",
    "\n",
    "X_hm_molecules.merge(\n",
    "    y_hm_molecules, \n",
    "    left_index=True, \n",
    "    right_index=True\n",
    ").to_excel('output/all__harmonic_mean__molecules_cnv.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Harmonic mean genomic variable.\n",
    "X_hm_molecules, y_hm_molecules = combine_tsv_files(\n",
    "    \"output/all__difference__No. Mutant Molecules per mL.tsv\",\n",
    "    \"output/all__difference__CNV Score.tsv\",\n",
    ")\n",
    "\n",
    "X_hm_molecules.merge(\n",
    "    y_hm_molecules, \n",
    "    left_index=True, \n",
    "    right_index=True\n",
    ").to_excel('output/all__difference__molecules_cnv.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Harmonic mean genomic variable.\n",
    "X_hm_molecules, y_hm_molecules = combine_tsv_files(\n",
    "    \"output/all__relative_difference__No. Mutant Molecules per mL.tsv\",\n",
    "    \"output/all__relative_difference__CNV Score.tsv\",\n",
    ")\n",
    "\n",
    "X_hm_molecules.merge(\n",
    "    y_hm_molecules, \n",
    "    left_index=True, \n",
    "    right_index=True\n",
    ").to_excel('output/all__relative_difference__molecules_cnv.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
